src/main/java/sg/ncl/MainController.java

```diff
@@ -471,10 +477,117 @@ public String loginSubmit(
 
     }
 
-    @RequestMapping("/passwordreset")
-    public String passwordreset(Model model) {
-        model.addAttribute("loginForm", new LoginForm());
-        return "passwordreset";
+    // triggered when user clicks "Forget Password?"
+    @RequestMapping("/password_reset_email")
+    public String passwordResetEmail(Model model) {
+        model.addAttribute("passwordResetRequestForm", new PasswordResetRequestForm());
+        return "password_reset_email";
+    }
+
+    // triggered when user clicks "Send Reset Email" button
+    @PostMapping("/password_reset_request")
+    public String sendPasswordResetRequest(
+            @ModelAttribute("passwordResetRequestForm") PasswordResetRequestForm passwordResetRequestForm
+    ) throws WebServiceRuntimeException
+    {
+        String email = passwordResetRequestForm.getEmail();
+        if(!VALID_EMAIL_ADDRESS_REGEX.matcher(email).matches()) {
+            passwordResetRequestForm.setErrMsg("Please provide a valid email address");
+            return "password_reset_email";
+        }
+
+        JSONObject obj = new JSONObject();
+        obj.put("username", email);
+
+        HttpHeaders headers = new HttpHeaders();
+        headers.setContentType(MediaType.APPLICATION_JSON);
+        HttpEntity<String> request = new HttpEntity<>(obj.toString(), headers);
+        restTemplate.setErrorHandler(new MyResponseErrorHandler());
+        ResponseEntity response = null;
+        try {
+            response = restTemplate.exchange(properties.getPasswordResetRequestURI(), HttpMethod.POST, request, String.class);
+        } catch (RestClientException e) {
+            log.warn("Error connecting to sio for password reset request!");
+            passwordResetRequestForm.setErrMsg("Cannot connect. Server may be down!");
+            return "password_reset_email";
+        }
+
+        if (RestUtil.isError(response.getStatusCode())) {
+            log.warn("server responded error for password reset request");
+            passwordResetRequestForm.setErrMsg("Email not registered. Please use a different email address.");
+            return "password_reset_email";
+        }
+
+        return "password_reset_email_sent";
+    }
+
+    // triggered when user clicks password reset link in the email
+    @RequestMapping(path = "/passwordReset", params = {"key"})
+    public String passwordResetNewPassword(@NotNull @RequestParam("key") final String key, Model model, HttpSession session) {
+        model.addAttribute("passwordResetForm", new PasswordResetForm());
+        session.setAttribute("key", key);
+        // redirect to the page for user to enter new password
+        return "password_reset_new_password";
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Define a constant instead of duplicating this literal "password_reset_new_password" 5 times. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS1192)